collector/lib/ResourceSelector.cpp

```diff
@@ -0,0 +1,96 @@
+//
+
+#include "ResourceSelector.h"
+
+#include <regex>
+
+#include "Hash.h"
+
+namespace collector {
+
+bool ResourceSelector::IsRuleValueFollowed(const storage::RuleValue& value, const std::string& resource_name) {
+  if (value.match_type() == storage::REGEX) {
+    std::regex pattern(value.value());
+    return std::regex_match(resource_name, pattern);
+  } else {
+    return (resource_name == value.value());
+  }
+}
+
+bool ResourceSelector::IsRuleFollowed(const storage::SelectorRule& rule, const std::string& resource_name) {
+  if (rule.operator_() == storage::BooleanOperator::OR) {
+    for (const auto& value : rule.values()) {
+      if (IsRuleValueFollowed(value, resource_name)) {
+        return true;
+      }
+    }
+  } else if (rule.operator_() == storage::BooleanOperator::AND) {
+    for (const auto& value : rule.values()) {
+      if (!IsRuleValueFollowed(value, resource_name)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  return false;
+}
+
+bool ResourceSelector::IsResourceInResourceSelector(const storage::ResourceSelector& rs, const std::string& resource_type, const std::string& resource_name) {
+  for (const auto& rule : rs.rules()) {
+    if (rule.field_name() == resource_type && IsRuleFollowed(rule, resource_name)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool ResourceSelector::IsNamespaceInResourceSelector(const storage::ResourceSelector& rs, const std::string& ns) {
+  return IsResourceInResourceSelector(rs, "Namespace", ns);
+}
+
+bool ResourceSelector::IsClusterInResourceSelector(const storage::ResourceSelector& rs, const std::string& cluster) {
+  return IsResourceInResourceSelector(rs, "Cluster", cluster);
+}
+
+bool ResourceSelector::IsNamespaceSelected(const storage::ResourceCollection& rc, const std::string& ns) {
+  for (const auto& rs : rc.resource_selectors()) {
+    if (IsNamespaceInResourceSelector(rs, ns)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool ResourceSelector::AreClusterAndNamespaceSelected(const storage::ResourceCollection& rc, const std::string& cluster, const std::string& ns) {
+  for (const auto& rs : rc.resource_selectors()) {
+    if (IsClusterInResourceSelector(rs, cluster) && IsNamespaceInResourceSelector(rs, ns)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool ResourceSelector::AreClusterAndNamespaceSelected(const storage::ResourceCollection& rc, const UnorderedMap<std::string, storage::ResourceCollection> rcMap, const std::string& cluster, const std::string& ns) {
+  if (AreClusterAndNamespaceSelected(rc, cluster, ns)) {
+    return true;
+  }
+
+  for (const auto& embeddedCollection : rc.embedded_collections()) {
+    auto embeddedRc = rcMap.find(embeddedCollection.id());
+    if (embeddedRc != rcMap.end()) {
+      bool inEmbeddedRc = AreClusterAndNamespaceSelected(embeddedRc->second, rcMap, cluster, ns);
+      if (inEmbeddedRc) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
```

Comment:
Improvements to this function include checks for cycles and stopping at a maximum depth.